peps/pep-0724.rst

```diff
@@ -0,0 +1,305 @@
+
+    def func1(val: A):
+        if guard(val):
+            reveal_type(val)  # NP
+        else:
+            reveal_type(val)  # NN
+
+Each use of a ``TypeGuard`` involves five types:
+
+* P = ``TypeGuard`` parameter type
+* R = ``TypeGuard`` return type
+* A = Argument type (pre-narrowed)
+* NP = Narrowed type (positive)
+* NN = Narrowed type (negative)
+
+This PEP proposes some modifications to :pep:`647` in order to allow a type
+checker to further narrow what a ``TypeGuard`` returns when the ``R`` type is
+consistent [#isconsistent]_ with the ``P`` type.
+
+* When the output type ``R`` of a ``TypeGuard`` is consistent [#isconsistent]_
+  with the type of its first input parameter (``P``), type checkers will apply
+  stricter type semantics:
+
+  .. code-block:: python
+
+     # Stricter TypeGuard possible (Kangaroo | Koala is consistent with Animal)
+     def is_marsupial(val: Animal) -> TypeGuard[Kangaroo | Koala]:
+         return isinstance(val, Kangaroo | Koala)
+
+     # Stricter TypeGuard not possible
+     # (list[T] is not consistent with list[T | None])
+     def has_no_nones(val: list[T | None]) -> TypeGuard[list[T]]:
+         return None not in val
+
+* When stricter semantics are possible, the output type for the ``False`` or
+  ``NN`` case can be determined:
+
+  .. code-block:: python
+
+     def is_str(val: str | int) -> TypeGuard[str]:  # Stricter mode possible
+         return isinstance(val, str)
+
+     def func(val: str | int):
+         if is_str(val):
+             # Type checkers can assume val is a 'str' in this branch
+         else:
+             # Type checkers can assume val is an 'int' in this branch
+
+* When the output type of the ``TypeGuard`` is a union, the type checker can
+  apply additional type narrowing based on the type of the first input
+  argument, eliminating union elements that are impossible given the input
+  argument type:
+
+  .. code-block:: python
+
+     def is_cardinal_direction(val: str) -> TypeGuard[Literal["N", "S", "E", "W"]]:
+         return val in ("N", "S", "E", "W")
+
+     def func(direction: Literal["NW", "E"]):
+         if is_cardinal_direction(direction):
+             # Literal["E"] The type cannot be "N", "S" or "W" here because of
+             # argument type
+         else:
+             # Literal["NW"]
+
+
+This logic can be summed up in a table:
+
+============ ======================= ===================
+\            Non strict TypeGuard    Strict TypeGuard
+============ ======================= ===================
+Applies when R not consistent with P R consistent with P
+NP is ..     :math:`R`               :math:`A \land R`
+NN is ..     :math:`A`               :math:`A \land \neg{R}`
+============ ======================= ===================
+
+Intersection of types and type negation are not defined and are left up to the
+type checker to decide on how to implement. Future extensions to the type
+system may define this behavior though.
```

Comment:
I was initially a little confused here, as this PEP doesn't propose ``Intersection[]`` or ``Negate[]``, but then I realised you probably mean achieving an intersection by using multiple type guards -- would it be worth adding a line of disambiguation?

Reply:
No, it doesn't mean achieving an intersection using multiple type guards. It means that strict type guards use intersection to define what they return. The return type of a strict type guard is the intersection of the input type (A) and the return type (R). This intersection isn't defined in Python yet.

Reply:
Pyright has its own implementation of this intersection at the moment.